sdk/storage/src/shares/share_file_client.cpp

```diff
@@ -0,0 +1,877 @@
+            {
+              ret = returnTypeConverter(chunk);
+            }
+          };
+
+    int64_t remainingOffset = firstChunkOffset + firstChunkLength;
+    int64_t remainingSize = fileRangeSize - firstChunkLength;
+    int64_t chunkSize;
+    if (options.ChunkSize.HasValue())
+    {
+      chunkSize = options.ChunkSize.GetValue();
+    }
+    else
+    {
+      int64_t c_grainSize = 4 * 1024;
+      chunkSize = remainingSize / options.Concurrency;
+      chunkSize = (std::max(chunkSize, int64_t(1)) + c_grainSize - 1) / c_grainSize * c_grainSize;
+      chunkSize = std::min(chunkSize, Storage::Details::c_DownloadDefaultChunkSize);
+    }
+
+    Storage::Details::ConcurrentTransfer(
+        remainingOffset, remainingSize, chunkSize, options.Concurrency, downloadChunkFunc);
+    ret->ContentLength = fileRangeSize;
+    return ret;
+  }
+
+  Azure::Core::Response<UploadFileFromResult> FileClient::UploadFrom(
+      const uint8_t* buffer,
+      std::size_t bufferSize,
+      const UploadFileFromOptions& options) const
+  {
+    ShareRestClient::File::CreateOptions protocolLayerOptions;
+    protocolLayerOptions.XMsContentLength = bufferSize;
+    protocolLayerOptions.FileAttributes = FileAttributesToString(options.SmbProperties.Attributes);
+    if (protocolLayerOptions.FileAttributes.empty())
+    {
+      protocolLayerOptions.FileAttributes = FileAttributesToString(FileAttributes::None);
+    }
+    protocolLayerOptions.FileCreationTime = options.SmbProperties.FileCreationTime;
+    protocolLayerOptions.FileLastWriteTime = options.SmbProperties.FileLastWriteTime;
+    if (options.FilePermission.HasValue())
+    {
+      protocolLayerOptions.FilePermission = options.FilePermission.GetValue();
+    }
+    else if (!options.SmbProperties.FilePermissionKey.empty())
+    {
+      protocolLayerOptions.FilePermissionKey = options.SmbProperties.FilePermissionKey;
+    }
+    else
+    {
+      protocolLayerOptions.FilePermission = "inherit";
+    }
+    if (options.HttpHeaders.HasValue())
+    {
+      if (!options.HttpHeaders.GetValue().ContentType.empty())
+      {
+        protocolLayerOptions.FileContentType = options.HttpHeaders.GetValue().ContentType;
+      }
+      if (!options.HttpHeaders.GetValue().ContentEncoding.empty())
+      {
+        protocolLayerOptions.FileContentEncoding = options.HttpHeaders.GetValue().ContentEncoding;
+      }
+      if (!options.HttpHeaders.GetValue().ContentLanguage.empty())
+      {
+        protocolLayerOptions.FileContentLanguage = options.HttpHeaders.GetValue().ContentLanguage;
+      }
+      if (!options.HttpHeaders.GetValue().CacheControl.empty())
+      {
+        protocolLayerOptions.FileCacheControl = options.HttpHeaders.GetValue().CacheControl;
+      }
+      if (!options.HttpHeaders.GetValue().ContentDisposition.empty())
+      {
+        protocolLayerOptions.FileContentDisposition
+            = options.HttpHeaders.GetValue().ContentDisposition;
+      }
+    }
```

Comment:
why do you have to check `HasValue()` and `.empty()` here?

Reply:
Even if `HttpHeaders` is not null, some of the headers can still be empty, this check avoids sending them as empty headers.

Reply:
1. The `HttpHeaders` is non-null. This is a conclusion from the private review meeting with Jeffrey. If you are not okay with this, bring it up and we can revisit this conclusion. I don't think silently ignore a conclusion is a good idea.

2. I don't think these headers should be nullable in protocol layer. But since we don't have any regulation about protocol layer, it's totally up to you how to define them. My suggestion is to make these non-nullable and add this HTTP header if it's not empty. In this way, your convenience layer code will be much simpler. But it's still totally up to you whether or not take this suggestion.

Reply:
I think I'm going to make HttpHeaders non-null, but keep protocol layer headers nullable for those headers that can be empty.